include/proxy-wasm/context_interface.h

```diff
@@ -0,0 +1,622 @@
+   * shutting down.
+   * @return true for stream contexts or for Root Context(s) if the VM can shutdown, false for Root
+   * Context(s) if the VM should wait until the Root Context calls the proxy_done() ABI call.  Note:
+   * the VM may (optionally) shutdown after some configured timeout even if the Root Context does
+   * not call proxy_done().
+   */
+  virtual bool onDone() = 0;
+
+  /**
+   * Call when no further stream calls will occur.  This will cause the corresponding Context in the
+   * VM to be deleted.
+   * Called by the host code.
+   */
+  virtual void onDelete();
+};
+
+/**
+ * HttpInterface is the interface between the VM host and the VM for HTTP streams. Note: headers and
+ * trailers are obtained by the code in the VM via the XXXHeaderMapXXX set of functions. Body data
+ * is obtained by the code in the VM via calls implemented by the proxy-independent host code using
+ * the getBuffer() call.
+ */
+struct HttpInterface {
+public:
+  /**
+   * Call on a stream context to indicate that the request headers have arrived.  Calls
+   * onCreate() to create a Context in the VM first.
+   */
+  virtual ProxyAction onRequestHeaders() = 0;
+
+  // Call on a stream context to indicate that body data has arrived.
+  virtual ProxyAction onRequestBody() = 0;
+
+  // Call on a stream context to indicate that the request trailers have arrived.
+  virtual ProxyAction onRequestTrailers() = 0;
+
+  // Call on a stream context to indicate that the request metadata has arrived.
+  virtual ProxyAction onRequestMetadata() = 0;
+
+  // Call on a stream context to indicate that the request trailers have arrived.
+  virtual ProxyAction onResponseHeaders() = 0;
+
+  // Call on a stream context to indicate that body data has arrived.
+  virtual ProxyAction onResponseBody() = 0;
+
+  // Call on a stream context to indicate that the request trailers have arrived.
+  virtual ProxyAction onResponseTrailers() = 0;
+
+  // Call on a stream context to indicate that the request metadata has arrived.
+  virtual ProxyAction onResponseMetadata() = 0;
+
+  /**
+   * Respond directly to an HTTP request.
+   * @param response_code is the response code to send.
+   * @param body is the body of the response.
+   * @param additional_headers are additional headers to send in the response.
+   * @param grpc_status is an optional gRPC status if the connection is a gRPC connection.
+   * @param details are details of any (gRPC) error.
+   */
+  virtual WasmResult sendLocalResponse(uint64_t response_code, string_view body,
+                                       Pairs additional_headers, uint64_t grpc_status,
+                                       string_view details) = 0;
+
+  // Call when the stream closes. See RootInterface.
+  virtual bool onDone() = 0;
+
+  // Call when the stream status has finalized, e.g. for logging. See RootInterface.
+  virtual void onFinalized() = 0;
```

Comment:
qq: what is the difference between onDone() and onFinalized()? It's not completely clear what the lifecycle difference is. Also compared to onDelete(). clarify?

Reply:
+1 I have the same question